### ISSUE TITLE
SDPO (#343, 2601.20802): GRPO baseline is DrGRPO without KL anchor — verl deep-dive

### ISSUE THREAD

**Issue by Proteusiq** [Proteusiq/unthinking#88]
Spinning out a methodological finding from deep-diving the [verl](https://github.com/verl-project/verl) framework that [lasgroup/SDPO](https://github.com/lasgroup/SDPO) is forked from. Companion to [#85](https://github.com/Proteusiq/unthinking/issues/85) (Trilogy triage) and [#86](https://github.com/Proteusiq/unthinking/issues/86) (SDFT reproducibility).

## TL;DR

\`\`\`
┌─────────────────────────────────────────────────────────────────────┐
│  SDPO's "GRPO baseline" is DrGRPO without KL regularization.        │
│  SDPO simultaneously adds an EMA teacher.                           │
│                                                                     │
│  COMPARISON THE PAPER MAKES:                                        │
│    SDPO (EMA-anchored DrGRPO + dense distillation)                  │
│      vs                                                             │
│    "GRPO" (DrGRPO + no KL loss + tiny batches, deliberately         │
│              unregularized)                                         │
│                                                                     │
│  SOME OF SDPO'S WIN COULD BE THE EMA TEACHER restoring trust-       │
│  region stabilization that the baseline deliberately removes —      │
│  not the rich-feedback signal per se.                               │
└─────────────────────────────────────────────────────────────────────┘
\`\`\`

This does not invalidate the paper, but it does mean the headline numbers (LCBv6 27.9 → 48.8 on Qwen3-8B) compare against a deliberately weak baseline that the verl docs themselves only recommend for length-debiasing experiments, not as a strong default.

### 1. SDPO is a soft fork of verl, not a recipe

\`lasgroup/SDPO\` vendors verl in-tree at \`SDPO/verl/...\`. PR #42 ('Wip external verl') attempted to externalize it but was closed unmerged. SDPO is therefore frozen against an early-2026 verl snapshot (last push 2026-02-18).

### 2. Files SDPO adds on top of verl

| File | Addition |
|---|---|
| \`verl/trainer/ppo/core_algos.py\` | \`compute_self_distillation_loss()\` lines 1085–1188 |
| \`verl/trainer/ppo/ray_trainer.py\` | \`_maybe_build_self_distillation_batch()\` lines 672–796 |
| \`verl/workers/actor/dp_actor.py\` | \`_update_teacher()\` (EMA) at 132–151; SDPO branch at 808–849 |
| \`verl/workers/config/actor.py\` | new \`SelfDistillationConfig\` dataclass |
| \`verl/trainer/config/sdpo.yaml\` | top-level Hydra config |
| \`verl/trainer/config/baseline_grpo.yaml\` | the GRPO baseline |

Notably, SDPO is **not** registered with \`@register_policy_loss('sdpo')\`. Instead, \`dp_actor.py\` hard-codes a branch \`if self_distillation_enabled: ... else: get_policy_loss_fn(loss_mode)\`. Deliberately bypasses verl's plugin API.

### 3. The 'GRPO baseline' is actually DrGRPO without KL loss

\`run_local_grpo.sh\` invokes \`--config-name baseline_grpo\`. That YAML overrides:

\`\`\`yaml
algorithm:
  adv_estimator: grpo
  norm_adv_by_std_in_grpo: False        # DrGRPO
  rollout_correction:
    rollout_is: token
    rollout_is_threshold: 2.0
\`\`\`

Plus \`use_kl_loss: False\` (the actor's default). Per verl's own \`docs/algo/grpo.md\`: 'DrGRPO ... \`algorithm.norm_adv_by_std_in_grpo: False\`, which turns off standard deviation norm.'

Compared to canonical verl GRPO (\`examples/grpo_trainer/run_qwen2-7b_seq_balance.sh\`, which scores 89.0/89.8 on GSM8k):
- canonical: \`train_batch_size=1024\`, \`rollout_n=5\`, \`kl_loss_coef=0.001\`, \`use_kl_loss=True\`, \`norm_adv_by_std_in_grpo=True\`
- SDPO baseline: \`train_batch_size=32\`, \`rollout_n=8\`, no KL loss, no advantage normalization

This is a **drastically smaller, KL-free, DrGRPO-style** baseline.

### 4. The JSD loss

\`verl/trainer/ppo/core_algos.py:1085–1188\`, function \`compute_self_distillation_loss\`:

\`\`\`python
# alpha=0.5 -> Generalized Jensen-Shannon Divergence
if self_distillation_config.alpha == 0.0:
    kl_loss = F.kl_div(student_distill_log_probs, teacher_distill_log_probs,
                       reduction='none', log_target=True)        # forward KL
elif self_distillation_config.alpha == 1.0:
    kl_loss = F.kl_div(teacher_distill_log_probs, student_distill_log_probs,
                       reduction='none', log_target=True)        # reverse KL
else:
    alpha = torch.tensor(self_distillation_config.alpha, ...)
    mixture_log_probs = torch.logsumexp(
        torch.stack([student_distill_log_probs + torch.log(1 - alpha),
                     teacher_distill_log_probs + torch.log(alpha)]), dim=0)
    kl_teacher = F.kl_div(mixture_log_probs, teacher_distill_log_probs, ..., log_target=True)
    kl_student = F.kl_div(mixture_log_probs, student_distill_log_probs, ..., log_target=True)
    kl_loss = torch.lerp(kl_student, kl_teacher, alpha)  # Generalized JSD
\`\`\`

\`run_local_sdpo.sh\` sets \`ALPHA=0.5\` → Jensen-Shannon Divergence, *not* the reverse KL the paper figures suggest.

### 5. The 'successful sibling as feedback' trick

\`verl/trainer/ppo/ray_trainer.py:710–745\`. The teacher's prompt is built by literally inlining the sibling rollout's full text via the template \`\"\\n\\nCorrect solution:\\n\\n{successful_previous_attempt}\\n\"\`. The teacher then re-encodes the *same student tokens* \`responses\` under that augmented prompt; the KL is between \`(student | original prompt)\` and \`(teacher | original prompt + sibling's full successful trajectory)\`.

**The 'feedback' isn't analyzed feedback — it's literally a sibling's complete answer pasted into the prompt as a cheat sheet.** The student is pushed toward whatever next-token distribution the same model produces *when shown that cheat sheet*. This is amortization, not reasoning. Strong predictive-thesis evidence.

### 6. \`dont_reprompt_on_self_success=True\` (default)

\`ray_trainer.py:650–669\`:
\`\`\`python
solution_idxs = success_by_uid[uid]
if dont_reprompt_on_self_success:
    solution_idxs = [j for j in solution_idxs if j != idx]
if len(solution_idxs) == 0: return None
solution_idx = solution_idxs[0]
\`\`\`

A successful sample is excluded as its own teacher. If a sample is the only successful one in its group, \`solution=None\` → \`self_distillation_mask=0\` → loss masked out. **Successful loners get no training signal**. Mitigates a degenerate fixed point but means SDPO requires ≥2 sibling successes per prompt to learn from binary-reward environments.

### 7. Algorithmic critiques in the SDPO issue tracker

| Issue | Status | Concern |
|---|---|---|
| [#18](https://github.com/lasgroup/SDPO/issues/18) | closed | Notation inconsistency in gradient estimator (App. B.1); author acknowledged equation as wrong |
| [#25](https://github.com/lasgroup/SDPO/issues/25) | closed | Reproduction matches at step 0–400 (~68% ToolUse), then collapses to ~0% by step ~500 |
| [#26](https://github.com/lasgroup/SDPO/issues/26) | open | SDPO degrades through training on Math (Qwen2.5-3B) |
| [#31](https://github.com/lasgroup/SDPO/issues/31) | open | Cannot determine which seed controls 3-seed std-error reported in paper |
| [#38](https://github.com/lasgroup/SDPO/issues/38) | open | grad_norm collapses to 1e-5 while loss looks normal — vanishing gradient (paper Fig. 18 shows 0–20) |
| [#41](https://github.com/lasgroup/SDPO/issues/41) | open | \`run_local_sdpo.sh\` defaults to temperature=0; with rollout.n=8 all siblings nearly identical, neutralizes GRPO variance-reduction premise |

The cluster #25, #26, #38 collectively suggests **high hyperparameter sensitivity / instability** that is not reflected in the paper's std-error bands.

## Effect on stance

\`\`\`
┌─────────────────────────────────────────────────────────────────────┐
│  STANCE: SUPPORTS (unchanged)                                       │
├─────────────────────────────────────────────────────────────────────┤
│                                                                     │
│  The mechanistic claim (teacher = student = same network; only      │
│  conditioning context differs) is unaffected by baseline tuning.    │
│                                                                     │
│  In fact the verl-extracted detail STRENGTHENS the predictive       │
│  reading:                                                           │
│                                                                     │
│  • 'Feedback' is literally a sibling's full answer pasted into the  │
│    prompt as a cheat sheet — pure conditional next-token prediction │
│  • The model isn't analyzing failures; it's imitating its own       │
│    cheat-sheet-conditioned output                                   │
│  • The EMA teacher provides implicit trust-region regularization;   │
│    the algorithm needs an anchor exactly because the underlying     │
│    'self-correction' is fragile                                     │
│                                                                     │
│  The CAVEAT applies to MAGNITUDES, not direction. Headline LCBv6    │
│  27.9 → 48.8 on Qwen3-8B is best read as 'max achievable across     │
│  hyperparameter / checkpoint sweep' rather than a typical run.      │
│                                                                     │
└─────────────────────────────────────────────────────────────────────┘
\`\`\`

## Action items

- [x] verl-context block added to \`analysis/explored/340-349/343_sdpo_rl_self_distillation.md\`
- [x] Reproducibility caveats section linked to specific GitHub issues
- [x] \`analysis/paper_graph.md\` updated with verl methodological note
- [ ] If anyone reproduces SDPO with \`use_kl_loss=True\` baseline (canonical verl GRPO) and the gap shrinks, that would be decisive — log it here

## Related

- [#85](https://github.com/Proteusiq/unthinking/issues/85) — Self-Distillation Trilogy original triage
- [#86](https://github.com/Proteusiq/unthinking/issues/86) — SDFT reproducibility caveats
- [#87](https://github.com/Proteusiq/unthinking/issues/87) — Stale counters across analysis docs

**Comment by joejajo** [comment]
Hi @Proteusiq 
Thanks for the detailed anaylsis 
Im trying to implement https://arxiv.org/pdf/2504.20571 on SDPO , and i get the math 500 baseline lower than the baseline , since you are expert will one shot traning work for this SDPO method ?

**Comment by Proteusiq** [comment]
All action items completed in commit 3c7e6035 and 0a2559e2:

- [x] verl-context block in analysis #343 (lines 22-27, 107-109)
- [x] Reproducibility caveats (GitHub issues #18/#25/#26/#31/#38/#41 documented)
- [x] paper_graph.md updated with verl methodological note (line 1480)
- [ ] Watch for canonical-GRPO reproduction (ongoing — not closeable)

Stance unchanged (supports). The verl deep-dive **strengthened** the predictive reading: 'feedback' is literally a sibling rollout pasted as a cheat sheet. Magnitude caveats apply to headline numbers, not mechanism.
